### ISSUE TITLE
[PROPOSAL] add hasUrl to HTML

### ISSUE THREAD

**Issue by lukasoppermann** [spatie/laravel-menu#11]
Hey,

I think it would make sense to add the `hasUrl` interface to `HTML` objects as well. This would mean they can be automatically activated by the menu. Realistically every menu item that is not a simple link will be some form of `HTML` or `HTML`-like (#9) object. This however does not mean that those objects to not have a link. They probably do and need to be activated just like link objects.

The main issue is, that this requires a change in the Interface, as it means that a `$url` with a value of `null` must be accepted, to indicate no link is present (for cases where adding html does not mean adding a link).

When `null` is present the interface would need to understand that it should not do the link stuff.

Once this is taken care of, implementing it should be very easy.

To try it out I extended the class like this. I needed to _fake_ a bit and use the string 'null'. To get it to work. However this would actually activate any that is supposed to not have a link, when the url `/null` is active.

```php
class spatieHTML extends HTML implements HasUrl{

     protected $prefixes = [];
     protected $url = [];

    protected function __construct(string $html, string $url = null)
    {
        $this->html = $html;
        $this->url = $url;
        $this->active = false;
        $this->initializeParentAttributes();
    }

    public static function raw(string $html, string $url = 'null')
    {
        return new static($html, $url);
    }

    public function getUrl(): string
   {
       if (empty($this->prefixes)) {
           return $this->url;
       }
       return implode('', $this->prefixes) . '/' . ltrim($this->url, '/');
   }

   public function segment(int $index)
   {
       $path = parse_url($this->url)['path'] ?? '';
       $segments = array_values(
           array_filter(
               explode('/', $path),
               function ($value) {
                   return $value !== '';
               }
           )
       );
       return $segments[$index - 1] ?? null;
   }
   /**
    * @param string $prefix
    *
    * @return $this
    */
   public function prefix(string $prefix)
   {
       $this->prefixes[] = $prefix;
       return $this;
   }

}
```

Now using it would be pretty easy:

```php
Menu::new()
    ->add(spatieHTML::raw(view('menu.item', $item)->render()), '/link')
    ->add(spatieHTML::raw(view('menu.notAnItem', $item)->render()));

```

Additionally the `prefixLinks` method would have to be adapted to not typehint `Link`. Alternatively a new method like `prefixItemLinks` could be created, that is basically a copy of `prefixLinks` just without the typehint.

**Comment by sebastiandedeyne** [comment]
I understand the concern, but I don't agree that the `Html` class should implement the `HasUrl` interface (I often use it for some raw html that isn't necessarily tied to a url).

I just pushed some changes that should make it easier to add your own implementation though.

- I created a `HasUrl` trait, so you don't need to reimplement the functionality
- I deprecated `prefixLinks` (it'll be removed in the next major version), and added a `prefixUrls` method that checks for the `HasUrl` interface instead

I wouldn't recommend extending `Html` for this though, since it's pretty annoying that you're constrained by method parameters. If you `use ActivatableTrait, HasUrl, ParentAttributes;` you've already covered most of the functionality though, so it's pretty trivial to reimplement everything instead of extending.

I'm going to look in to the `addView` issue now, and tag a new version with the above changes afterwards.
